### PR TITLE
8271054: [REDO] Wrong stage gets focused after modal stage creation

### PR DESCRIPTION
**Pull request by tsayao**
Found the problem thru this path:

**WindowStage.java**
```
    final void handleFocusDisabled() {
        if (activeWindows.isEmpty()) {
            return;
        }
        WindowStage window = activeWindows.get(activeWindows.size() - 1);
        window.setIconified(false);
        window.requestToFront();
        window.requestFocus();
    }
```

**glass_window.cpp**
```
void WindowContextBase::process_focus(GdkEventFocus* event) {
   ...

    if (jwindow) {
        if (!event->in || isEnabled()) {
            mainEnv->CallVoidMethod(jwindow, jWindowNotifyFocus,
                    event->in ? com_sun_glass_events_WindowEvent_FOCUS_GAINED : com_sun_glass_events_WindowEvent_FOCUS_LOST);
            CHECK_JNI_EXCEPTION(mainEnv)
        } else {
            mainEnv->CallVoidMethod(jwindow, jWindowNotifyFocusDisabled);
            CHECK_JNI_EXCEPTION(mainEnv)
        }
    }
}
```

So `glass_window.cpp` was triggering `jWindowNotifyFocusDisabled` which triggered the java code on the Primary Stage (on the bug reproduce code).

The docs states:
```
    /**
     * Enables or disables the window.
     *
     * A disabled window is unfocusable by definition.
     * Also, key or mouse events aren't generated for disabled windows.
     *
     * When a user tries to activate a disabled window, or the window gets
     * accidentally brought to the top of the stacking order, the window
     * generates the FOCUS_DISABLED window event. A Glass client should react
     * to this event and bring the currently active modal blocker of the
     * disabled window to top by calling blocker's minimize(false), toFront(),
     * and requestFocus() methods. It may also 'blink' the blocker window to
     * further attract user's attention.
     *
     .....
```

So I guess the C++ side looks ok, java side on `handleFocusDisabled` I did not understand why it `requestToFront` and `requestFocus` on the latest window.

The solution makes disabled windows unfocusable and prevents mouse and keyboard events as the docs states, making it compatible with other platforms.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271054](https://bugs.openjdk.java.net/browse/JDK-8271054): [REDO] Wrong stage gets focused after modal stage creation

### Reviewers
 * [Pankaj Bansal](https://openjdk.java.net/census#pbansal) (@pankaj-bansal - Committer) 🔄 Re-review required (review applies to eb93b8b950f548f405bce4626abceb4e77c2e80e)
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/598/head:pull/598` \
`$ git checkout pull/598`

Update a local copy of the PR: \
`$ git checkout pull/598` \
`$ git pull https://git.openjdk.java.net/jfx pull/598/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 598`

View PR using the GUI difftool: \
`$ git pr show -t 598`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/598.diff">https://git.openjdk.java.net/jfx/pull/598.diff</a>

</details>

### CODE REVIEW

**Comment by tsayao** [comment]
image

**Comment by pankaj-bansal** [comment]
I ran the testcase attached in JBS with the fix multiple times. I can still reproduce the issue on Ubuntu 20.0 almost always, so the fix does not seem to work for me.

**Comment by tsayao** [comment]
Weird, It works consistently for me on 20.04. Just tested again to be sure.

**Comment by pankaj-bansal** [comment]
> Weird, It works consistently for me on 20.04. Just tested again to be sure.

I am running a 20.04 VM. The test fails for me 60-70% of the time. I will request someone in team to try this once.

**Comment by tsayao** [comment]
@pankaj-bansal Could you please re-test to confirm that it does not work on your VM?

**Comment by kevinrushforth** [comment]
I just tested this on my Ubuntu 20.04 VM and the problem still reproduces for me with or without this fix.

**Comment by tsayao** [comment]
Docs states:
```java
        /*
         * When a user tries to activate a disabled window, or the window gets
         * accidentally brought to the top of the stacking order, the window
         * generates the FOCUS_DISABLED window event. A Glass client should react
         * to this event and bring the currently active modal blocker of the
         * disabled window to top by calling blocker's minimize(false), toFront(),
         * and requestFocus() methods. It may also 'blink' the blocker window to
         * further attract user's attention.
         */
```

But `WindowStage.java` actually does:

```java
    final void handleFocusDisabled() {
        if (activeWindows.isEmpty()) {
            return;
        }
        WindowStage window = activeWindows.get(activeWindows.size() - 1);
        window.setIconified(false);
        window.requestToFront();
        window.requestFocus();
    }
```
It brings up current activeWindows.size() - 1, not actually the MODAL blocker. Seems wrong to me.

The event is born on `glass_window.cpp`:

```c++
mainEnv->CallVoidMethod(jwindow, jWindowNotifyFocusDisabled);
```

The current proposed code (which I will redo) blocks this line from being sent (on non-virtual machines apparently). But commenting the `jWindowNotifyFocusDisabled` should make the problem disappear (just to test the theory).

**Comment by tsayao** [comment]
I have proposed a different fix, but it might have concurrency issues and probably others that should be investigated.

**Comment by tsayao** [comment]
Just a note, our javafx application at the company I work for is unusable with javafx17 on Linux, windows go behind their owner or pops up wrongly. Can't use 17 line until it's fixed.

**Comment by tsayao** [comment]
@pankaj-bansal Can you try the test attached to the JBS on macos and windows? Thanks.

**Comment by tsayao** [comment]
The problem occurs because `WindowStage.activeWindows` gets out of order, because of focus events. So when "Focus Disabled" happens it brings up the wrong window.

Probably because "Focus Stealing" situation which is mentioned on:
https://docs.gtk.org/gtk3/method.Window.present_with_time.html

I'm working on the fix.

**Comment by tsayao** [comment]
The ideal fix would be to know the original event and save it's X11 serial event number to pass to `gtk_window_present_with_time` so "focus stealing" would not happen. But I don't think it's possible because it should save the "event serial number" on the original mouse click to open a window (for example). As you have thought, many "original event" combinations are possible.

The fix just reorders the `WindowStage.activeWindows` by calling `FOCUS_GAINED`. So when `FOCUS_DISABLED` happens it focuses the correct window. Please not that with this fix there are extra calls to FOCUS events and reorder of `WindowStage.activeWindows` (although they are not expensive).

**Comment by tsayao** [comment]
It would be nice to have this fix on jfx 17, since currently the issue causes random window focused and it's very confusing to the end user.

**Comment by pankaj-bansal** [comment]
I ran the full systemTest and this test is not introducing any regression. Also, the manual test attached in JBS works fine with this change.

File: modules/javafx.graphics/src/main/native-glass/gtk/GlassApplication.cpp [hunk 1]
```diff
@@ -445,6 +445,7 @@ static void process_events(GdkEvent* event, gpointer data)
     }
 
     if (ctx != NULL) {
+
```
Comment (pankaj-bansal) [inline modules/javafx.graphics/src/main/native-glass/gtk/GlassApplication.cpp:448]: You can probably remove this unnecessary change as this is only one change in the file.

**Comment by kevinrushforth** [comment]
On my Ubuntu 20.04 VM the bug still happens about 1/2 the time with this fix.

**Comment by tsayao** [comment]
@kevinrushforth can you try this lastest version, please?

**Comment by kevinrushforth** [comment]
I'll review and test it on a couple different systems.

**Review by kevinrushforth — APPROVED** [review]
This fix looks good. The test program for this bug reliably works now, and I verified that it doesn't reintroduce [JDK-8240640](https://bugs.openjdk.java.net/browse/JDK-8240640) or [JDK-8227366](https://bugs.openjdk.java.net/browse/JDK-8227366)

File: modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp [hunk 2]
```diff
@@ -775,6 +776,8 @@ WindowContextTop::WindowContextTop(jobject _jwindow, WindowContext* _owner, long
     if (frame_type == TITLED) {
         request_frame_extents();
     }
+
+    event_serial = 0;
```
Comment (arapte) [inline modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp:780]: should we use GDK_CURRENT_TIME instead of 0 ?
GDK_CURRENT_TIME is defined also as 0 and represent current time.

File: modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp [hunk 3]
```diff
@@ -1371,7 +1384,10 @@ void WindowContextTop::request_focus() {
     //The WindowContextBase::set_visible will take care of showing the window.
     //The below code will only handle later request_focus.
     if (is_visible()) {
-        gtk_window_present(GTK_WINDOW(gtk_widget));
+        // event_serial holds an event serial that will help X11 determine which window should have the focus and
+        // prevents activeWindows on WindowStage.java to be out of order which may cause the FOCUS_DISABLED event
+        // to bring up the wrong window (it brings up the last which will not be the real "last" if out of order).
+        gtk_window_present_with_time(GTK_WINDOW(gtk_widget), event_serial);
```
Comment (arapte) [inline modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp:1390]: `event_serial` is not reset after use. I could observe that a stale value of `event_serial` gets reused several times.
Steps to observe the behavior:
1. Print event_serial in this method
2. Run the sample program attached to JBS.
3. Press the "Click Me!" button 
4. Answer YES on the Alert 
5. Three stages should be visible on the screen.
6. Relocate such that all stages are visible
7. Click on 'This should be on Top' Stage
8. Click on "This is a stage with no owner' stage.
9. Click on StageTest icon in Taskbar. Keep repeating this step. The stage gains and looses focus and the same event_serial gets printed on terminal.
=> It does not look harmful behavior wise. But can observe that `event_serial` gets reused.
Screenshot: 
Screenshot 2022-03-08 at 6 53 46 PM



Will it be good idea to reset it to 0/GDK_CURRENT_TIME and use something like,

```
if (event_serial == GDK_CURRENT_TIME) {
    gtk_window_present(GTK_WINDOW(gtk_widget));
} else {
    gtk_window_present_with_time(GTK_WINDOW(gtk_widget), event_serial);
    event_serial = GDK_CURRENT_TIME;
}
```

OR

```
gtk_window_present_with_time(GTK_WINDOW(gtk_widget), event_serial);
event_serial = GDK_CURRENT_TIME;
```

**Review by arapte — COMMENTED** [review]
The fix itself looks good. though I have a query. I did not observe any mis-behaviour, so please check if the suggested change is required or not.

File: modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp (hunk 2 above)
Comment (tsayao) [reply modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp]: GDK_CURRENT_TIME is always  0, but it looks better. Changed it.

File: modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp (hunk 3 above)
Comment (tsayao) [reply modules/javafx.graphics/src/main/native-glass/gtk/glass_window.cpp:1390]: You're right, it can avoid other possible errors.

**Review by kevinrushforth — APPROVED** [review]

**Review by arapte — APPROVED** [review]
